### PR TITLE
Feature/trip profiling

### PR DESCRIPTION
**Pull request by countcain**

### CODE REVIEW

File: ios/RNSentiance.m [hunk 1]
```diff
@@ -711,5 +743,58 @@ - (NSDictionary*)convertCrashEventToDict:(NSDate*)date lastKnownLocation:(CLLoca
...
+            double startTime = [transportSegment.startDate timeIntervalSince1970] * 1000;
+            transportSegmentDict[@"startTime"] = @(startTime);
+            double endTime = [transportSegment.endDate timeIntervalSince1970] * 1000;
+            transportSegmentDict[@"endTime"] = @(endTime);
+            transportSegmentDict[@"distance"] = @(transportSegment.distance);
+            transportSegmentDict[@"averageSpeed"] = @(transportSegment.averageSpeed);
+            transportSegmentDict[@"topSpeed"] = @(transportSegment.topSpeed);
+            transportSegmentDict[@"percentOfTimeSpeeding"] = @(transportSegment.speedingPercentage);
+            SENTTripProcessingVehicleMode vehicleMode = transportSegment.vehicleMode;
+            switch (vehicleMode) {
```
Comment (sebouh00) [inline ios/RNSentiance.m:766]: Nit: Similar to the java counterpart, moving this into its own method would simplify this method.

File: android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java [hunk 2]
```diff
@@ -208,4 +214,60 @@ public static String convertUserActivityType(UserActivityType activityType) {
...
+          WritableMap transportSegmentMap = Arguments.createMap();
+          transportSegmentMap.putDouble("startTime", (double) transportSegment.getStartTime());
+          transportSegmentMap.putDouble("endTime", (double) transportSegment.getEndTime());
+          transportSegmentMap.putDouble("distance", transportSegment.getDistance());
+          transportSegmentMap.putDouble("averageSpeed", transportSegment.getAverageSpeed());
+          transportSegmentMap.putDouble("topSpeed", transportSegment.getTopSpeed());
+          transportSegmentMap.putInt("percentOfTimeSpeeding", transportSegment.getPercentOfTimeSpeeding());
+          VehicleMode vehicleMode = transportSegment.getVehicleMode();
+          String mode;
+          switch(vehicleMode) {
```
[the remark this answers is not in the document]
Comment (countcain) [reply android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java]: Moved to its own method.

File: android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java [hunk 3]
```diff
@@ -208,4 +214,60 @@ public static String convertUserActivityType(UserActivityType activityType) {
...
   }
 
+  public static WritableMap convertTripProfile(TripProfile tripProfile) {
+    WritableMap map = Arguments.createMap();
+
+    try {
+      map.putString("tripId", tripProfile.getTripId());
+      List<TransportSegment> transportSegments = tripProfile.getTransportSegments();
+      WritableArray transportSegmentsArray = Arguments.createArray();
+      if (!transportSegments.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (countcain) [reply android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java]: Removed.

File: ios/RNSentiance.m (hunk 1 above)
Comment (countcain) [reply ios/RNSentiance.m]: Moved to its own method.

File: android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java [hunk 4]
```diff
@@ -208,4 +214,60 @@ public static String convertUserActivityType(UserActivityType activityType) {
...
+              mode = "NOT_VEHICLE";
+              break;
+            default:
+              mode = "UNKNOWN";
+          }
+          transportSegmentMap.putString("vehicleMode", mode);
+
+          List<HardEvent> hardEvents = transportSegment.getHardEvents();
+          WritableArray hardEventsArray = Arguments.createArray();
+          if (!hardEvents.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (countcain) [reply android/src/main/java/com/sentiance/react/bridge/RNSentianceConverter.java]: I noticed that hardEvents is nullable. So I added a null check here. Same for `getDistance()` `getAverageSpeed()` `getTopSpeed()` `getPercentOfTimeSpeeding()`

File: ios/RNSentiance.m [hunk 5]
```diff
@@ -514,6 +514,38 @@ - (NSString *) getValueForKey:(NSString *)key value:(NSString *)defaultValue {
...
+    }];
+}
+
+RCT_EXPORT_METHOD(updateTripProfileConfig:(NSDictionary *)config
+                  resolver:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject)
+{
+    @try {
+        if (config == nil || config[@"enableFullProfiling"] == nil) {
+            @throw([NSException exceptionWithName:@"NilException" reason:@"enableFullProfiling is not provided" userInfo:nil]);
```
Comment (ralf601) [inline ios/RNSentiance.m:534]: Shouldn't we reject promise instead of throwing exception?

Comment (countcain) [reply ios/RNSentiance.m]: Updated. fyi. @throw functions the same as reject()  in RCT_EXPORT_METHOD while I was testing. The js catch block will be invoked with same param not matter it is from @throw or reject(). Not sure about if there are differences in other aspects.

File: ios/RNSentiance.m [hunk 6]
```diff
@@ -514,6 +514,39 @@ - (NSString *) getValueForKey:(NSString *)key value:(NSString *)defaultValue {
...
+    }];
+}
+
+RCT_EXPORT_METHOD(updateTripProfileConfig:(NSDictionary *)config
+                  resolver:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject)
+{
+    @try {
+        if (config == nil || config[@"enableFullProfiling"] == nil) {
+            reject(@"NilException", @"enableFullProfiling is not provided", nil);
```
Comment (ralf601) [inline ios/RNSentiance.m:534]: rejection should have same error for both platforms, like on android `E_SDK_MISSING_PARAMS` is used.

**Pull request merged**